x/oracle/keeper/keeper.go

```diff
@@ -183,6 +160,87 @@ func (k Keeper) SetFeeInfo(ctx sdk.Context, networkDescriptor types.NetworkDescr
 	return nil
 }
 
+// SetProphecyWithInitValue set the prophecy in keeper
+func (k Keeper) SetProphecyWithInitValue(ctx sdk.Context, prophecyID []byte) {
+	prophecy := types.Prophecy{
+		Id:              prophecyID,
+		Status:          types.StatusText_STATUS_TEXT_PENDING,
+		ClaimValidators: []string{},
+	}
+	k.SetProphecy(ctx, prophecy)
+}
+
+// ProcessSignProphecy deal with the signature from validator
+func (k Keeper) ProcessSignProphecy(ctx sdk.Context, networkDescriptor types.NetworkDescriptor, prophecyID []byte, cosmosSender, ethereumAddress, signature string) error {
+	prophecy, ok := k.GetProphecy(ctx, prophecyID)
+	if !ok {
+		return types.ErrProphecyNotFound
+	}
+
+	whiteList := k.GetOracleWhiteList(ctx, types.NewNetworkIdentity(networkDescriptor))
+	power, ok := whiteList.WhiteList[cosmosSender]
+	if !ok {
+		return errors.New("message sender to sign prophecy not in the whitelist")
+	}
+
+	if power == 0 {
+		return errors.New("message sender to sign prophecy without vote power")
+	}
+
+	// verify the signature
+	publicKey, err := gethCrypto.Ecrecover(prophecyID, gethCommon.FromHex(signature))
+	if err != nil {
+		return err
+	}
+
+	ok = gethCrypto.VerifySignature(publicKey, prophecyID, []byte(signature))
+	if !ok {
+		return errors.New("incorrect signature")
```

Comment:
Let's change this error message to `incorrect ethereum signature` so that it is a more descriptive error.

Reply:
accepted